utils/blackbox.py

```diff
@@ -0,0 +1,348 @@
+
+
+def blackbox_pyteal(subr: SubroutineFnWrapper, mode: Mode) -> Callable[..., Expr]:
+    """Functor producing ready-to-compile PyTeal programs from annotated subroutines
+
+    Args:
+        subr: annotated subroutine to wrap inside program.
+            Note: the `input_types` parameters should be supplied to @Subroutine() annotation
+        mode: type of program to produce: logic sig (Mode.Signature) or app (Mode.Application)
+
+    Returns:
+        a function that called with no parameters -e.g. result()-
+        returns a PyTeal expression compiling to a ready-to-test TEAL program.
+
+    The return type is callable in order to adhere to the API of end-to-end unit tests.
+
+    Generated TEAL code depends on the mode, subroutine input types, and subroutine output types.
+    * logic sigs:
+        * input received via `arg i`
+        * args are converted (cf. "input conversion" below) and passed to the subroutine
+        * subroutine output is not logged (log is not available)
+        * subroutine output is converted (cf "output conversion" below)
+    * apps:
+        * input received via `txna ApplicationArgs i`
+        * args are converted (cf. "input conversion" below) and passed to the subroutine
+        * subroutine output is logged after possible conversion (cf. "logging conversion")
+        * subroutine output is converted (cf "output conversion" below)
+    * input conversion:
+        * Empty input array:
+            do not read any args and call subroutine immediately
+        * arg of TealType.bytes and TealType.anytype:
+            read arg and pass to subroutine as is
+        * arg of TealType.uint64:
+            convert arg to int using Btoi() when received
+        * pass-by-ref ScratchVar arguments:
+            in addition to the above -
+                o store the arg (or converted arg) in a ScratchVar
+                o invoke the subroutine using this ScratchVar instead of the arg (or converted arg)
+    * output conversion:
+        * TealType.uint64:
+            provide subroutine's result to the top of the stack when exiting program
+        * TealType.bytes:
+            convert subroutine's result to the top of the stack to its length and then exit
+        * TealType.none or TealType.anytype:
+            push Int(1337) to the stack as it is either impossible (TealType.none),
+            or unknown at compile time (TealType.anytype) to convert to an Int
+    * logging conversion:
+        * TealType.uint64:
+            convert subroutine's output using Itob() and log the result
+        * TealType.bytes:
+            log the subroutine's result
+        * TealType.none or TealType.anytype:
+            log Itob(Int(1337)) as it is either impossible (TealType.none),
+            or unknown at compile time (TealType.anytype) how to convert to Bytes
+
+
+    Example 1: Using blackbox_pyteal for a simple test of both an app and logic sig:
```

Comment:
@jasonpaulos @michaeldiamant - where should this doc live? I could see a new section `PyTeal Dryrun Testing` coming after `Compiler Options`. Also, this doesn't seem to properly belong under `API` and I would argue that we should leave that section untouched.

Reply:
@tzaffi A few thoughts:
* I agree with a new section under _Compiler Options_.  I'd name it _Blackbox testing_ or similar with _blackbox_ to stick with the naming applied throughout the repo.
* Additional thoughts:
  * Are we also saying the inline doc examples highlighted by the thread will be removed?
    * Open to discussion:  My preference is _yes_ because the examples aren't compiled and will go stale.  As an example, I've already opened several, minor spelling fix _suggestions_ in 2 places.
   * Open to discussion:  I'll follow the group's preference, but want to note that I'd be on-board to agree on _where_ user guides will be documented here.  And then supply the docs in a new PR because there's already a lot happening here. 
* Aside:  The question makes me wonder if a broader user guide re-organization makes sense.  For the moment, I'm thinking we opt for a change specific to the problem at-hand.

Reply:
Coming back to the thread with fresher eyes, I'm changing my position a bit.  Happy to discuss live if easier.

I think my rubric is to favor:
* Accelerate merging the PR.
* Retain control to make backwards _incompatible_ changes to blackbox infrastructure.  At least until we gain a bit more experience.

My view is that blackbox testing is _definitely_ useful for PyTeal developers.  A primary question in my mind is:  What level of blackbox testing exposure (if any) do we provide to developers?

Given that rubric, I'm open to a few options for the PR's purpose:
* Opt to make blackbox testing a tool internal to PyTeal maintainers - If so, I'd _probably_ lean towards _no_ user guide docs because the user guide doesn't cover maintainers today (possibly, a separate concern).  I think we can provide inline pointers to examples and/or other lightweight references.
* Opt to make blackbox testing available to PyTeal developers - If so:
  * Declare at least in comments that the tooling is liable to change.  
  * I'd be open to keeping the docs as discussed in the prior note.

Reply:
committed to link in comments to where tests actually run.

Reply:
Here is the new version. I've refactored the `user_guide_test` and removed the long executable portion of the `blackbox_pyteal()`'s comment:
https://github.com/algorand/pyteal/blob/6afe1af7400cc12a7aeaaf70cc6faaa5009d8db9/utils/blackbox.py#L105

Reply:
@tzaffi Thanks for the updates - I consider the thread resolved.  Since I didn't open it, I leave it open for you to resolve in case you're holding for other feedback.

Summarizing changes for anyone else looking in:
* We opted for _no_ user guide docs in favor of an inline doc pointer from `blackbox_pyteal` to examples.
* We retained `user_guide_test.py` as a unit test.  It had temporarily been moved under _integration_ prior to last changeset.